envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt

```diff
@@ -222,18 +240,153 @@ class RBACFilterFactory(
+            listOf(
+                Principal.newBuilder().setMetadata(
+                    MetadataMatcher.newBuilder()
+                        .setFilter("envoy.filters.http.header_to_metadata")
+                        .addPath(
+                            MetadataMatcher.PathSegment.newBuilder().setKey("jwt-status").build()
+                        )
+                        .setValue(
+                            ValueMatcher.newBuilder().setStringMatch(StringMatcher.newBuilder().setExact("present"))
+                        )
+                ).build(),
+                Principal.newBuilder().setMetadata(
+                    MetadataMatcher.newBuilder()
+                        .setFilter("envoy.filters.http.jwt_authn")
+                        .addPath(
+                            MetadataMatcher.PathSegment.newBuilder()
+                                .setKey(jwtProperties.payloadInMetadata)
+                        ).addPath(
+                            MetadataMatcher.PathSegment.newBuilder()
+                                .setKey(jwtProperties.fieldRequiredInToken)
+                        )
+                        .setValue(ValueMatcher.newBuilder().setPresentMatch(true)).build()
+                ).build()
+            )
+        ).build()
+    ).build()
+
+    private val allowMissingPolicyPrincipal = Principal.newBuilder().setOrIds(
+        Principal.Set.newBuilder().addAllIds(
+            listOf(
+                Principal.newBuilder()
+                    .setMetadata(
+                        MetadataMatcher.newBuilder()
+                            .setFilter("envoy.filters.http.header_to_metadata")
+                            .addPath(
+                                MetadataMatcher.PathSegment.newBuilder().setKey("jwt-status").build()
+                            )
+                            .setValue(
+                                ValueMatcher.newBuilder().setStringMatch(StringMatcher.newBuilder().setExact("missing"))
+                            )
+                    )
+                    .build(),
+                strictPolicyPrincipal
+            )
+        ).build()
+    ).build()
+
+    private fun jwtClientWithSelectorPrincipal(client: ClientWithSelector, oAuthProvider: OAuthProvider): Principal {
+        return Principal.newBuilder().setMetadata(
+            MetadataMatcher.newBuilder()
+                .setFilter("envoy.filters.http.jwt_authn")
+                .addPath(
+                    MetadataMatcher.PathSegment.newBuilder()
+                        .setKey(jwtProperties.payloadInMetadata).build()
+                )
+                .addPath(
+                    MetadataMatcher.PathSegment.newBuilder()
+                        .setKey(oAuthProvider.selectorToTokenField[client.selector]).build()
+                )
+                .setValue(
+                    ValueMatcher.newBuilder().setListMatch(
+                        ListMatcher.newBuilder().setOneOf(
```

Comment:
Why do we need `ListMatcher` if there is only one value?

Reply:
Field in token which we are checking can be a List, so it won't work with `ValueMatcher` only